include/boost/fusion/container/vector/vector.hpp

```diff
@@ -9,12 +9,620 @@
+        {
+            typedef vector_data<T...> base;
+
+            BOOST_STATIC_ASSERT((base::size::value == N));
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data()
+            {}
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data(vector_data const& rhs)
+                : base(static_cast<base const &>(rhs))
+            {}
+
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data(vector_data&& rhs)
+                : base(std::forward<base>(rhs))
+            {}
+
+            template <typename ...U>
+            BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data(U&&... var)
+                : base(std::forward<U>(var)...)
+            {}
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data &
+            operator=(vector_data const& rhs)
+            {
+                base::operator=(static_cast<base const&>(rhs));
+                return *this;
+            }
+
+            BOOST_CXX14_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+            vector_data &
+            operator=(vector_data&& rhs)
+            {
+                base::operator=(std::forward<base>(rhs));
+                return *this;
+            }
+
+            using base::operator=;
+        };
+
+        template <typename D, typename... T>
+        struct construct_vector_;
+
+        template <typename D>
+        struct construct_vector_<D> { typedef D type; };
+
+        template <typename... U, typename... Tail>
+        struct construct_vector_<vector_data<U...>, void_, Tail...>
+            : construct_vector_<vector_data<U...> > {};
+
+        template <typename... U, typename Head, typename... Tail>
+        struct construct_vector_<vector_data<U...>, Head, Tail...>
+            : construct_vector_<vector_data<U..., Head>, Tail...> {};
+
+        template <typename... T>
+        struct construct_vector : construct_vector_<vector_detail::vector_data<>, T...> {};
+    }
+
+    // This class provides backward compatibility: vector<T, ..., void_, void_, ...>.
+    template <typename... T>
+    struct vector
+        : vector_detail::construct_vector<T...>::type
+    {
+        typedef typename vector_detail::construct_vector<T...>::type base;
+
+        BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+        vector()
+        {}
+
+        BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+        vector(vector const& rhs)
+            : base(static_cast<base const&>(rhs))
+        {}
+
+        BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+        vector(vector&& rhs)
+            : base(std::forward<base>(rhs))
+        {}
+
+        // rvalue-references is required here in order to forward any arguments to
+        // base: vector(T const&...) doesn't work with trailing void_ and
+        // vector(U const&...) cannot forward any arguments to base.
+        template <typename... U>
+        BOOST_FUSION_GPU_ENABLED
+        vector(U&&... u)
```

Comment:
Is [this problem](https://akrzemi1.wordpress.com/2013/10/10/too-perfect-forwarding/) an issue here? The only difference are the variadics.